{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d8256254-c868-4ac4-b1db-dda611434e4d",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "Attaching sp\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n",
      "Loading required package: pkgmaker\n",
      "\n",
      "Loading required package: registry\n",
      "\n",
      "Loading required package: rngtools\n",
      "\n",
      "Loading required package: cluster\n",
      "\n",
      "NMF - BioConductor layer [OK] | Shared memory capabilities [NO: bigmemory] | Cores 31/32\n",
      "\n",
      "  To enable shared memory capabilities, try: install.extras('\n",
      "NMF\n",
      "')\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.4.2\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "  genomic data. Bioinformatics 2016.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: igraph\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:NMF’:\n",
      "\n",
      "    algorithm, compare\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    normalize, path, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.15\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘circlize’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:igraph’:\n",
      "\n",
      "    degree\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘pracma’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expm, lu, tril, triu\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "library(NMF)\n",
    "library(ggalluvial)\n",
    "library(ComplexHeatmap)\n",
    "library(CellChat)\n",
    "library(patchwork)\n",
    "library(graphics)\n",
    "library(Matrix)\n",
    "options(stringsAsFactors = FALSE)\n",
    "\n",
    "library(circlize)\n",
    "library(colorspace)\n",
    "options(repr.plot.width = 12, repr.plot.height = 9, repr.plot.res = 300)\n",
    "\n",
    "library(pracma)\n",
    "library(glue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d291f5b3-f527-4840-bdcc-d2f38bda4c94",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An object of class CellChat created from a single dataset \n",
      " 34361 genes.\n",
      " 218999 cells.An object of class CellChat created from a single dataset \n",
      " 34361 genes.\n",
      " 182306 cells."
     ]
    }
   ],
   "source": [
    "# # for plotting figA, one C2C network\n",
    "# cellchat <- readRDS(file = \"/gpfs/gibbs/pi/girgenti/JZhang/CL/cellChat/brain_call/new/SZBD-Kellis_annotated-CON_cellchat.rds\")\n",
    "\n",
    "# for plotting figC, differential network\n",
    "disease = \"BD\"\n",
    "\n",
    "cellchat.DIS <- readRDS(file = glue(\"/gpfs/gibbs/pi/girgenti/JZhang/CL/cellChat/brain_call/new/SZBD-Kellis_annotated-{disease}_cellchat.rds\"))\n",
    "cellchat.CT <- readRDS(file = \"/gpfs/gibbs/pi/girgenti/JZhang/CL/cellChat/brain_call/new/SZBD-Kellis_annotated-CON_cellchat.rds\")\n",
    "# object.list <- list(CT = cellchat.CT, DIS = cellchat.DIS)\n",
    "# object.list <- list(DIS = cellchat.DIS, CON = cellchat.CT)\n",
    "# cellchat <- mergeCellChat(object.list, add.names = names(object.list))\n",
    "# cellchat\n",
    "\n",
    "print(cellchat.CT)\n",
    "print(cellchat.DIS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d39201b6-5e3a-42dd-bafc-04a3cf5a480e",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_name = c(\"Astro\",\"Endo\",\"Immune\",\"Micro\",\"Oligo\",\"OPC\",\"PC\",\"SMC\",\"VLMC\",\n",
    "                  \"Chandelier\",\"Lamp5\",\"Lamp5 Lhx6\",\"Pax6\",\"Pvalb\",\"Sncg\",\"Sst\", \"Sst Chodl\",\"Vip\",\n",
    "                  \"L2/3 IT\",\"L4 IT\",\"L5 ET\",\"L5 IT\",\"L5/6 NP\",\"L6 CT\",\"L6 IT\",\"L6 IT Car3\",\"L6b\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3646b939-3773-44ce-b658-14c33f888a9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "inhibitory_celltypes = c(\"Chandelier\",\"Lamp5\",\"Lamp5 Lhx6\",\"Pax6\",\"Pvalb\",\"Sncg\",\"Sst\", \"Sst Chodl\",\"Vip\")\n",
    "excitatory_celltypes = c(\"L2/3 IT\",\"L4 IT\",\"L5 ET\",\"L5 IT\",\"L5/6 NP\",\"L6 CT\",\"L6 IT\",\"L6 IT Car3\",\"L6b\")\n",
    "support_celltypes = c(\"Astro\",\"Endo\",\"Immune\",\"Micro\",\"Oligo\",\"OPC\",\"PC\",\"SMC\",\"VLMC\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "018306d9-7f22-4294-9934-95edc4d4274d",
   "metadata": {},
   "source": [
    "## Fig B: NMF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "098dc9cc-31fa-48ba-b33c-28e58eea90c0",
   "metadata": {},
   "source": [
    "### Global Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f051751a-c85f-4bbe-aea0-121b9ab6ca7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#versions that will be made for each combination (resulting in that many figures being made)\n",
    "versions = seq(from = 1, to = 1, by = 1) \n",
    "# NMF methods to be used, can be \"brunet\" or \"lee\"...\n",
    "methods = list(\"brunet\")\n",
    "# NMF seeding methods to be used, can be \"random\" or \"nndsvd\"...\n",
    "seeds = list(\"random\")\n",
    "# NMF number of runs to be used, can be 200, 500, 1000, 2000, 5000, 10000...\n",
    "nrums = list(200)\n",
    "\n",
    "# these are manually ligand ordering, if left bank it will be done automatically using patterns from control\n",
    "incoming_ligand_order <- c(\n",
    "    \"ANGPTL\", \"BMP\", \"COMPLEMENT\", \"CXCL\", \"EDN\", \"ENHO\", \"GALECTIN\", \"GRN\", \"MIF\", \"ncWNT\", \"PSAP\", \"PTN\", \"SPP1\", \"TGFb\", \"VEGF\", \"VISFATIN\", \n",
    "    \"CCK\", \"CRH\", \"GDF\", \"IGF\", \"KIT\", \"NPY\", \"PTH\", \"SOMATOSTATIN\", \"TAC\", \"VIP\", \n",
    "    \"ANGPT\", \"CSF\", \"CX3C\", \"EGF\", \"FGF\", \"GAS\", \"HGF\", \"NRG\", \"NT\", \"PACAP\", \"PARs\", \"PDGF\", \"PERIOSTIN\", \"PROS\", \"SEMA3\", \"UCN\", \"WNT\")\n",
    "outgoing_ligand_order <- c(\n",
    "    \"ANGPTL\", \"BMP\", \"COMPLEMENT\", \"CXCL\", \"EDN\", \"ENHO\", \"GALECTIN\", \"GRN\", \"MIF\", \"ncWNT\", \"PTN\", \"SPP1\", \"TGFb\", \"VEGF\", \"VISFATIN\", \n",
    "    \"CCK\", \"CRH\", \"GDF\", \"IGF\", \"KIT\", \"NPY\", \"PTH\", \"SOMATOSTATIN\", \"TAC\", \"VIP\", \n",
    "    \"ANGPT\", \"CSF\", \"CX3C\", \"EGF\", \"FGF\", \"GAS\", \"HGF\", \"NRG\", \"NT\", \"PACAP\", \"PARs\", \"PDGF\", \"PERIOSTIN\", \"PROS\", \"PSAP\", \"SEMA3\", \"UCN\", \"WNT\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d7ef39e7-16dc-4f7f-b490-445b3583189a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "43"
      ],
      "text/latex": [
       "43"
      ],
      "text/markdown": [
       "43"
      ],
      "text/plain": [
       "[1] 43"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(incoming_ligand_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ce9a0c1f-c749-4981-93f4-04d7f809c4da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "43"
      ],
      "text/latex": [
       "43"
      ],
      "text/markdown": [
       "43"
      ],
      "text/plain": [
       "[1] 43"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(outgoing_ligand_order)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff72a749-17c4-43b5-9217-7ad54d4290f0",
   "metadata": {},
   "source": [
    "### Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e03542de-4a2e-49d5-a9c3-f1adb16337ec",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"beginning NMF\"\n",
      "[1] \"Finished NMF\"\n",
      "[1] \"creating directory\"\n",
      " [1] \"FGF\"          \"NT\"           \"NRG\"          \"WNT\"          \"GAS\"         \n",
      " [6] \"PERIOSTIN\"    \"ANGPT\"        \"CX3C\"         \"PACAP\"        \"PROS\"        \n",
      "[11] \"PSAP\"         \"PARs\"         \"PDGF\"         \"EGF\"          \"UCN\"         \n",
      "[16] \"SEMA3\"        \"CSF\"          \"HGF\"          \"BMP\"          \"COMPLEMENT\"  \n",
      "[21] \"PTN\"          \"ncWNT\"        \"GALECTIN\"     \"GRN\"          \"VISFATIN\"    \n",
      "[26] \"SPP1\"         \"MIF\"          \"ANGPTL\"       \"VEGF\"         \"CXCL\"        \n",
      "[31] \"ENHO\"         \"TGFb\"         \"EDN\"          \"KIT\"          \"CCK\"         \n",
      "[36] \"VIP\"          \"IGF\"          \"NPY\"          \"TAC\"          \"PTH\"         \n",
      "[41] \"GDF\"          \"CRH\"          \"SOMATOSTATIN\"\n",
      "[1] 90\n",
      "[1] \"saving fig:\"\n",
      "[1] \"/gpfs/ysm/project/girgenti/cl2553/C2C/plots/NMF/control/brunet/random/brunet_random_outgoing200_runs_1.pdf\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"done saving fig\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Please make sure you have load `library(ggalluvial)` when running this function\n",
      "\n",
      "#f9918a#f9918a#f9918a#f9918a#f9918a#f9918a#f9918a#f9918a#f9918a#33c860#33c860#33c860#33c860#33c860#33c860#33c860#33c860#33c860#81b0ff#81b0ff#81b0ff#81b0ff#81b0ff#81b0ff#81b0ff#81b0ff#81b0ff#f9918a#33c860#81b0ff\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"saving fig:\"\n",
      "[1] \"/gpfs/ysm/project/girgenti/cl2553/C2C/plots/NMF/control/brunet/random/brunet_random_ORG_outgoing200_runs_1.pdf\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"done saving fig\"\n",
      "[1] \"making directory\"\n",
      " [1] \"KIT\"          \"CCK\"          \"VIP\"          \"IGF\"          \"NPY\"         \n",
      " [6] \"PTH\"          \"GDF\"          \"CRH\"          \"SOMATOSTATIN\" \"TAC\"         \n",
      "[11] \"FGF\"          \"NT\"           \"NRG\"          \"WNT\"          \"GAS\"         \n",
      "[16] \"PERIOSTIN\"    \"CX3C\"         \"PACAP\"        \"PROS\"         \"ANGPT\"       \n",
      "[21] \"PARs\"         \"PDGF\"         \"EGF\"          \"UCN\"          \"SEMA3\"       \n",
      "[26] \"CSF\"          \"HGF\"          \"BMP\"          \"COMPLEMENT\"   \"PTN\"         \n",
      "[31] \"ncWNT\"        \"GALECTIN\"     \"GRN\"          \"VISFATIN\"     \"SPP1\"        \n",
      "[36] \"PSAP\"         \"ANGPTL\"       \"VEGF\"         \"CXCL\"         \"ENHO\"        \n",
      "[41] \"MIF\"          \"TGFb\"         \"EDN\"         \n",
      "[1] 1360\n",
      "[1] \"saving fig:\"\n",
      "[1] \"/gpfs/ysm/project/girgenti/cl2553/C2C/plots/NMF/control/brunet/random/brunet_random_incoming200_runs_1.pdf\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"done saving fig\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Please make sure you have load `library(ggalluvial)` when running this function\n",
      "\n",
      "#c5b233#c5b233#c5b233#c5b233#c5b233#c5b233#c5b233#c5b233#c5b233#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#f783e9#f783e9#f783e9#f783e9#f783e9#f783e9#f783e9#f783e9#f783e9#c5b233#33ccd0#f783e9\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"saving fig:\"\n",
      "[1] \"/gpfs/ysm/project/girgenti/cl2553/C2C/plots/NMF/control/brunet/random/brunet_random_ORG_incoming200_runs_1.pdf\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"done saving fig\"\n",
      "[1] \"beginning NMF\"\n",
      "[1] \"Finished NMF\"\n",
      "[1] \"creating directory\"\n",
      " [1] \"BMP\"          \"PTN\"          \"MIF\"          \"SPP1\"         \"VEGF\"        \n",
      " [6] \"VISFATIN\"     \"TGFb\"         \"EDN\"          \"ANGPT\"        \"ENHO\"        \n",
      "[11] \"COMPLEMENT\"   \"ANGPTL\"       \"SEMA3\"        \"NT\"           \"PSAP\"        \n",
      "[16] \"PERIOSTIN\"    \"PDGF\"         \"NPY\"          \"GAS\"          \"EGF\"         \n",
      "[21] \"SOMATOSTATIN\" \"PARs\"         \"CX3C\"         \"WNT\"          \"FGF\"         \n",
      "[26] \"CSF\"          \"UCN\"          \"TAC\"          \"CCK\"          \"PROS\"        \n",
      "[31] \"VIP\"          \"IGF\"          \"NRG\"          \"KIT\"          \"CRH\"         \n",
      "[36] \"PTH\"         \n",
      "[1] 48\n",
      "[1] \"saving fig:\"\n",
      "[1] \"/gpfs/ysm/project/girgenti/cl2553/C2C/plots/NMF/BD/brunet/random/brunet_random_outgoing200_runs_1.pdf\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"done saving fig\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Please make sure you have load `library(ggalluvial)` when running this function\n",
      "\n",
      "#f9918a#f9918a#f9918a#f9918a#f9918a#f9918a#f9918a#f9918a#f9918a#33c860#33c860#33c860#33c860#33c860#33c860#33c860#33c860#33c860#33c860#81b0ff#81b0ff#81b0ff#81b0ff#81b0ff#81b0ff#81b0ff#81b0ff#f9918a#33c860#81b0ff\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"saving fig:\"\n",
      "[1] \"/gpfs/ysm/project/girgenti/cl2553/C2C/plots/NMF/BD/brunet/random/brunet_random_ORG_outgoing200_runs_1.pdf\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"done saving fig\"\n",
      "[1] \"making directory\"\n",
      " [1] \"BMP\"          \"PTN\"          \"MIF\"          \"SPP1\"         \"VEGF\"        \n",
      " [6] \"VISFATIN\"     \"TGFb\"         \"EDN\"          \"ANGPT\"        \"ENHO\"        \n",
      "[11] \"PSAP\"         \"ANGPTL\"       \"COMPLEMENT\"   \"SEMA3\"        \"NT\"          \n",
      "[16] \"PERIOSTIN\"    \"PDGF\"         \"NPY\"          \"GAS\"          \"EGF\"         \n",
      "[21] \"SOMATOSTATIN\" \"PARs\"         \"CX3C\"         \"WNT\"          \"FGF\"         \n",
      "[26] \"CSF\"          \"UCN\"          \"TAC\"          \"CCK\"          \"PROS\"        \n",
      "[31] \"IGF\"          \"NRG\"          \"KIT\"          \"VIP\"          \"PTH\"         \n",
      "[36] \"CRH\"         \n",
      "[1] 1560\n",
      "[1] \"saving fig:\"\n",
      "[1] \"/gpfs/ysm/project/girgenti/cl2553/C2C/plots/NMF/BD/brunet/random/brunet_random_incoming200_runs_1.pdf\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"done saving fig\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Please make sure you have load `library(ggalluvial)` when running this function\n",
      "\n",
      "#c5b233#c5b233#c5b233#c5b233#c5b233#c5b233#c5b233#c5b233#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#33ccd0#f783e9#f783e9#f783e9#f783e9#f783e9#f783e9#f783e9#f783e9#c5b233#33ccd0#f783e9\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"saving fig:\"\n",
      "[1] \"/gpfs/ysm/project/girgenti/cl2553/C2C/plots/NMF/BD/brunet/random/brunet_random_ORG_incoming200_runs_1.pdf\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"done saving fig\"\n"
     ]
    }
   ],
   "source": [
    "# --------------------------------------------\n",
    "# run the function\n",
    "# --------------------------------------------\n",
    "# !! IMPORTANT !!\n",
    "# RIGHT NOW FOR LIGAND ORDERING TO WORK YOU MUST RUN \n",
    "# CONTROL FIRST (i'm working on fixing this)\n",
    "# --------------------------------------------\n",
    "\n",
    "NMF_and_river(cellchat.CT,\"/gpfs/ysm/project/girgenti/cl2553/C2C/plots/NMF\", \"control\", inhibitory_celltypes, excitatory_celltypes, support_celltypes, parallel='p16')\n",
    "NMF_and_river(cellchat.DIS,\"/gpfs/ysm/project/girgenti/cl2553/C2C/plots/NMF\", disease, inhibitory_celltypes, excitatory_celltypes, support_celltypes, parallel='p16')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "138ae52a-f31f-4de1-b293-35e0c2720f3d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cfbbc97-6eb9-42fc-b815-d22ff873ced5",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### my_netAnalysis_river"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e3375134-7015-4f93-b46c-0844830c9b5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ----------------------------\n",
    "# this is the function to make\n",
    "# We include this for comparision purposes\n",
    "# ----------------------------\n",
    "my_netAnalysis_river <- function(object, slot.name = \"netP\", pattern = c(\"outgoing\",\"incoming\"), cutoff = 0.5,\n",
    "                              sources.use = NULL, targets.use = NULL, signaling = NULL,\n",
    "                              color.use = NULL, color.use.pattern = NULL, color.use.signaling = \"#c5c7c9\",\n",
    "                              do.order = FALSE, main.title = NULL,\n",
    "                              font.size = 5, font.size.title = 12){\n",
    "  message(\"Please make sure you have load `library(ggalluvial)` when running this function\")\n",
    "  requireNamespace(\"ggalluvial\")\n",
    "  #  suppressMessages(require(ggalluvial))\n",
    "  res.pattern <- methods::slot(object, slot.name)$pattern[[pattern]]\n",
    "  data1 = res.pattern$pattern$cell\n",
    "  data2 = res.pattern$pattern$signaling\n",
    "  if (is.null(color.use.pattern)) {\n",
    "    nPatterns <- length(unique(data1$Pattern))\n",
    "    if (pattern == \"outgoing\") {\n",
    "      color.use.pattern = ggPalette(nPatterns*2)[seq(1,nPatterns*2, by = 2)]\n",
    "    } else if (pattern == \"incoming\") {\n",
    "      color.use.pattern = ggPalette(nPatterns*2)[seq(2,nPatterns*2, by = 2)]\n",
    "    }\n",
    "  }\n",
    "  if (is.null(main.title)) {\n",
    "    if (pattern == \"outgoing\") {\n",
    "      main.title = \"Outgoing communication patterns of secreting cells\"\n",
    "    } else if (pattern == \"incoming\") {\n",
    "      main.title = \"Incoming communication patterns of target cells\"\n",
    "    }\n",
    "  }\n",
    "\n",
    "  if (is.null(data2)) {\n",
    "    data1$Contribution[data1$Contribution < cutoff] <- 0\n",
    "    plot.data <- data1\n",
    "    nPatterns<-length(unique(plot.data$Pattern))\n",
    "    nCellGroup<-length(unique(plot.data$CellGroup))\n",
    "    if (is.null(color.use)) {\n",
    "      color.use <- scPalette(nCellGroup)\n",
    "    }\n",
    "    if (is.null(color.use.pattern)){\n",
    "      color.use.pattern <- ggPalette(nPatterns)\n",
    "    }\n",
    "\n",
    "    plot.data.long <- to_lodes_form(plot.data, axes = 1:2, id = \"connection\")\n",
    "    if (do.order) {\n",
    "      mat = tapply(plot.data[[\"Contribution\"]], list(plot.data[[\"CellGroup\"]], plot.data[[\"Pattern\"]]), sum)\n",
    "      d <- dist(as.matrix(mat))\n",
    "      hc <- hclust(d, \"ave\")\n",
    "      k <- length(unique(grep(\"Pattern\", plot.data.long$stratum[plot.data.long$Contribution != 0], value = T)))\n",
    "      cluster <- hc %>% cutree(k)\n",
    "      order.name <- order(cluster)\n",
    "      plot.data.long$stratum <- factor(plot.data.long$stratum, levels = c(names(cluster)[order.name], colnames(mat)))\n",
    "      color.use <- color.use[order.name]\n",
    "    }\n",
    "    color.use.all <- c(color.use, color.use.pattern)\n",
    "    gg <- ggplot(plot.data.long,aes(x = factor(x, levels = c(\"CellGroup\", \"Pattern\")),y=Contribution,\n",
    "                                    stratum = stratum, alluvium = connection,\n",
    "                                    fill = stratum, label = stratum)) +\n",
    "      geom_flow(width = 1/3,aes.flow = \"backward\") +\n",
    "      geom_stratum(width=1/3,size=0.1,color=\"black\", alpha = 0.8, linetype = 1) +\n",
    "      geom_text(stat = \"stratum\", size = font.size) +\n",
    "      scale_x_discrete(limits = c(),  labels=c(\"Cell groups\", \"Patterns\")) +\n",
    "      scale_fill_manual(values = alpha(color.use.all, alpha = 0.8), drop = FALSE) +\n",
    "      theme_bw()+\n",
    "      theme(legend.position = \"none\",\n",
    "            axis.title = element_blank(),\n",
    "            axis.text.y= element_blank(),\n",
    "            panel.grid.major = element_blank(),\n",
    "            panel.grid.minor  = element_blank(),\n",
    "            panel.border = element_blank(),\n",
    "            axis.ticks = element_blank(),axis.text=element_text(size=10))+\n",
    "      ggtitle(main.title)\n",
    "\n",
    "  } else {\n",
    "    data1$Contribution[data1$Contribution < cutoff] <- 0\n",
    "    plot.data <- data1\n",
    "    nPatterns<-length(unique(plot.data$Pattern))\n",
    "    nCellGroup<-length(unique(plot.data$CellGroup))\n",
    "    cells.level = levels(object@idents)\n",
    "    if (is.null(color.use)) {\n",
    "      color.use <- scPalette(length(cells.level))[cells.level %in% unique(plot.data$CellGroup)]\n",
    "    }\n",
    "    if (is.null(color.use.pattern)){\n",
    "      color.use.pattern <- ggPalette(nPatterns)\n",
    "    }\n",
    "    if (!is.null(sources.use)) {\n",
    "      if (is.numeric(sources.use)) {\n",
    "        sources.use <- cells.level[sources.use]\n",
    "      }\n",
    "      plot.data <- subset(plot.data, CellGroup %in% sources.use)\n",
    "    }\n",
    "    if (!is.null(targets.use)) {\n",
    "      if (is.numeric(targets.use)) {\n",
    "        targets.use <- cells.level[targets.use]\n",
    "      }\n",
    "      plot.data <- subset(plot.data, CellGroup %in% targets.use)\n",
    "    }\n",
    "    ## connect cell groups with patterns\n",
    "    plot.data.long <- to_lodes_form(plot.data, axes = 1:2, id = \"connection\")\n",
    "    if (do.order) {\n",
    "      mat = tapply(plot.data[[\"Contribution\"]], list(plot.data[[\"CellGroup\"]], plot.data[[\"Pattern\"]]), sum)\n",
    "      d <- dist(as.matrix(mat))\n",
    "      hc <- hclust(d, \"ave\")\n",
    "      k <- length(unique(grep(\"Pattern\", plot.data.long$stratum[plot.data.long$Contribution != 0], value = T)))\n",
    "      cluster <- hc %>% cutree(k)\n",
    "      order.name <- order(cluster)\n",
    "      plot.data.long$stratum <- factor(plot.data.long$stratum, levels = c(names(cluster)[order.name], colnames(mat)))\n",
    "      color.use <- color.use[order.name]\n",
    "    }\n",
    "\n",
    "\n",
    "    data_subset <- subset(data1, data1$Contribution > 0.4)\n",
    "\n",
    "    data_subset\n",
    "\n",
    "    data_subset[[\"Pattern\"]]\n",
    "    # length(patterns)\n",
    "\n",
    "    colors = list()\n",
    "\n",
    "\n",
    "    if (pattern == \"outgoing\") {\n",
    "        for (cell_type in data_subset[[\"Pattern\"]]){\n",
    "        if(cell_type == \"Pattern 1\"){\n",
    "        colors <- append(colors, \"#f9918a\")\n",
    "        }\n",
    "        if(cell_type == \"Pattern 2\"){\n",
    "        colors <- append(colors,\"#33c860\")\n",
    "        }\n",
    "        if(cell_type == \"Pattern 3\"){\n",
    "        colors <- append(colors,\"#81b0ff\")\n",
    "        }\n",
    "    }\n",
    "    colors <- append(colors, \"#f9918a\")\n",
    "    colors <- append(colors,\"#33c860\")\n",
    "    colors <- append(colors,\"#81b0ff\")\n",
    "    } \n",
    "    else if (pattern == \"incoming\") {\n",
    "        for (cell_type in data_subset[[\"Pattern\"]]){\n",
    "            if(cell_type == \"Pattern 1\"){\n",
    "                colors <- append(colors, \"#c5b233\")\n",
    "            }\n",
    "            if(cell_type == \"Pattern 2\"){\n",
    "                colors <- append(colors,\"#33ccd0\")\n",
    "            }\n",
    "            if(cell_type == \"Pattern 3\"){\n",
    "                colors <- append(colors,\"#f783e9\")\n",
    "            }\n",
    "        }\n",
    "        colors <- append(colors, \"#c5b233\")\n",
    "        colors <- append(colors,\"#33ccd0\")\n",
    "        colors <- append(colors,\"#f783e9\")\n",
    "    }\n",
    "\n",
    "\n",
    "    names(colors) <- data_subset[[\"CellGroup\"]]\n",
    "    names(colors)[length(colors)-2] <- \"Pattern 1\"\n",
    "    names(colors)[length(colors)-1] <- \"Pattern 2\"\n",
    "    names(colors)[length(colors)] <- \"Pattern 3\"\n",
    "    # color.use <- colors\n",
    "\n",
    "    color.use.all <- c(color.use,color.use.pattern)\n",
    "\n",
    "    color.use.all <- colors\n",
    "\n",
    "    message(color.use.all)\n",
    "\n",
    "    # color.use.all <- colors\n",
    "\n",
    "\n",
    "    StatStratum <- ggalluvial::StatStratum\n",
    "    gg1 <- ggplot(plot.data.long,aes(x = factor(x, levels = c(\"CellGroup\", \"Pattern\")),y=Contribution,\n",
    "                                     stratum = stratum, alluvium = connection,\n",
    "                                     fill = stratum, label = stratum)) +\n",
    "      geom_flow(width = 1/3,aes.flow = \"forward\") +\n",
    "      geom_stratum(width=1/3,size=0.1,color=\"black\", alpha = 0.8, linetype = 1) +\n",
    "      geom_text(stat = \"stratum\", size = font.size) +\n",
    "      scale_x_discrete(limits = c(),  labels=c(\"Cell groups\", \"Patterns\")) +\n",
    "\n",
    "      scale_fill_manual(values = alpha(color.use.all, alpha = 0.8), drop = FALSE) +\n",
    "      theme_bw()+\n",
    "      theme(legend.position = \"none\",\n",
    "            axis.title = element_blank(),\n",
    "            axis.text.y= element_blank(),\n",
    "            panel.grid.major = element_blank(),\n",
    "            panel.grid.minor  = element_blank(),\n",
    "            panel.border = element_blank(),\n",
    "            axis.ticks = element_blank(),axis.text=element_text(size=10)) +\n",
    "      theme(plot.margin = unit(c(0, 0, 0, 0), \"cm\"))\n",
    "    ## connect patterns with signaling\n",
    "    data2$Contribution[data2$Contribution < cutoff] <- 0\n",
    "    plot.data <- data2\n",
    "    nPatterns<-length(unique(plot.data$Pattern))\n",
    "    nSignaling<-length(unique(plot.data$Signaling))\n",
    "    if (length(color.use.signaling) == 1) {\n",
    "      color.use.all <- c(color.use.pattern, rep(color.use.signaling, nSignaling))\n",
    "    } else {\n",
    "      color.use.all <- c(color.use.pattern, color.use.signaling)\n",
    "    }\n",
    "\n",
    "    if (!is.null(signaling)) {\n",
    "      plot.data <- plot.data[plot.data$Signaling %in% signaling, ]\n",
    "    }\n",
    "\n",
    "    plot.data.long <- ggalluvial::to_lodes_form(plot.data, axes = 1:2, id = \"connection\")\n",
    "\n",
    "    gg2 <- ggplot(plot.data.long,aes(x = factor(x, levels = c(\"Pattern\", \"Signaling\")),y= Contribution,\n",
    "                                     stratum = stratum, alluvium = connection,\n",
    "                                     fill = stratum, label = stratum)) +\n",
    "      geom_flow(width = 1/3,aes.flow = \"forward\") +\n",
    "      geom_stratum(width=1/3,size=0.1,color=\"black\", alpha = 0.8, linetype = 1) +\n",
    "      geom_text(stat = \"stratum\", size = font.size) + # 2.5\n",
    "      scale_x_discrete(limits = c(),  labels=c(\"Patterns\", \"Signaling\")) +\n",
    "      scale_fill_manual(values = alpha(color.use.all, alpha = 0.8), drop = FALSE) +\n",
    "      theme_bw()+\n",
    "      theme(legend.position = \"none\",\n",
    "            axis.title = element_blank(),\n",
    "            axis.text.y= element_blank(),\n",
    "            panel.grid.major = element_blank(),\n",
    "            panel.grid.minor  = element_blank(),\n",
    "            panel.border = element_blank(),\n",
    "            axis.ticks = element_blank(),axis.text=element_text(size= 10))+\n",
    "      theme(plot.margin = unit(c(0, 0, 0, 0), \"cm\"))\n",
    "\n",
    "    gg <- cowplot::plot_grid(gg1, gg2,align = \"h\", nrow = 1)\n",
    "    title <- cowplot::ggdraw() + cowplot::draw_label(main.title,size = font.size.title)\n",
    "    gg <- cowplot::plot_grid(title, gg, ncol=1, rel_heights=c(0.1, 1))\n",
    "  }\n",
    "  return(gg)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de3bfb27-dbb0-4c05-9386-d8e708aab1b9",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### my_merged_river"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0adb5805-5729-4176-8719-2d3230fd23e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------------------------------------------\n",
    "# Custom merged figured.\n",
    "# This is what we will use on the file paper\n",
    "#--------------------------------------------\n",
    "my_merged_river <- function(object,pattern, disease, inhibitory_celltypes, excitatory_celltypes, support_celltypes){\n",
    "\n",
    "    if (pattern == \"outgoing\") {\n",
    "      main.title = \"Outgoing communication patterns of secreting cells\"\n",
    "      # The color palette is defined in the function ggPalette\n",
    "      # color.use just needs three colors you can define them in a list with hex codes if you want \n",
    "      color.use <- ggPalette(3*2)[seq(1,3*2, by = 2)]\n",
    "      \n",
    "    } else if (pattern == \"incoming\") {\n",
    "      main.title = \"Incoming communication patterns of target cells\"\n",
    "      # The color palette is defined in the function ggPalette\n",
    "      # color.use just needs three colors you can define them in a list with hex codes if you want \n",
    "      color.use <- ggPalette(3*2)[seq(2,3*2, by = 2)]\n",
    "    }\n",
    "\n",
    "    slot.name = \"netP\"\n",
    "    res.pattern <- methods::slot(object, slot.name)$pattern[[pattern]]\n",
    "\n",
    "    data1 = res.pattern$pattern$cell\n",
    "    data2 = res.pattern$pattern$signaling\n",
    "\n",
    "    sorted_group <- data2[order(data2$Signaling, -data2$Contribution),]\n",
    "    data2 <- sorted_group[!duplicated(sorted_group$Signaling),]\n",
    "\n",
    "    sorted_group <- data1[order(data1$CellGroup, -data1$Contribution),]\n",
    "    data1 <- sorted_group[!duplicated(sorted_group$CellGroup),]\n",
    "\n",
    "    plot.data <- merge(data1,data2,by=\"Pattern\")\n",
    "\n",
    "    plot.data$Signaling <- as.character(plot.data$Signaling)\n",
    "    print(unique(plot.data$Signaling))\n",
    "    plot.data <- plot.data[order(plot.data$Pattern, plot.data$Signaling),]\n",
    "\n",
    "    # -- this is code for pulling out custom ligands\n",
    "    # ligand_in_order = ligand_in_order[!ligand_in_order %in% c(\"FGF\", \"NRG\", \"PSAP\")]\n",
    "    # ligand_in_order = c(ligand_in_order, \"FGF\", \"NRG\", \"PSAP\")\n",
    "    # plot.data$Signaling <- factor(plot.data$Signaling, levels = ligand_in_order)\n",
    "\n",
    "    # order ligands by pattern\n",
    "    if(pattern == \"outgoing\"){\n",
    "      if(length(outgoing_ligand_order) > 0){\n",
    "          plot.data$Signaling <- factor(plot.data$Signaling, levels = outgoing_ligand_order) \n",
    "      } \n",
    "      else {\n",
    "        if(disease == \"control\"){\n",
    "            outgoing_ligand_order <<- c()\n",
    "            for (pattern in unique(plot.data$Pattern)){\n",
    "              outgoing_ligand_order <<- c(outgoing_ligand_order, unique(plot.data$Signaling[plot.data$Pattern == pattern]))\n",
    "            }\n",
    "            plot.data$Signaling <- factor(plot.data$Signaling, levels = outgoing_ligand_order) \n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    else if(pattern == \"incoming\"){\n",
    "      if(length(incoming_ligand_order) > 0){\n",
    "          plot.data$Signaling <- factor(plot.data$Signaling, levels = incoming_ligand_order) \n",
    "      } \n",
    "      else {\n",
    "        if(disease == \"control\"){\n",
    "            incoming_ligand_order <<- c()\n",
    "            for (pattern in unique(plot.data$Pattern)){\n",
    "              incoming_ligand_order <<- c(incoming_ligand_order, unique(plot.data$Signaling[plot.data$Pattern == pattern]))\n",
    "            }\n",
    "            plot.data$Signaling <- factor(plot.data$Signaling, levels = incoming_ligand_order) \n",
    "        }\n",
    "      }\n",
    "    }\n",
    "\n",
    "    #order cell types by cell type groups\n",
    "    # inhibitory_celltypes <- c(\"Chandelier\", \"Lamp5\", \"Lamp5 Lhx6\", \"Pax6\", \"Pvalb\", \"Sst\", \"Sst Chodl\", \"Vip\", \"Sncg\")\n",
    "    inhibitory_celltypes <- inhibitory_celltypes[order(inhibitory_celltypes)]\n",
    "    # excitatory_celltypes <- c(\"L2/3 IT\", \"L4 IT\", \"L5\", \"L5 ET\", \"L5/6 ET\", \"L6 CT\", \"L6 IT\", \"L6 IT Car3\",\"L5 IT\",\"L5/6 NP\",\"L6b\")\n",
    "    excitatory_celltypes <- excitatory_celltypes[order(excitatory_celltypes)]\n",
    "    # support_celltypes <- c(\"Astro\",\"Endo\",\"Micro/PVM\", \"Oligo\", \"OPC\", \"VLMC\")\n",
    "\n",
    "    celltypesOrder <- c(support_celltypes, inhibitory_celltypes, excitatory_celltypes)\n",
    "    plot.data$CellGroup <- factor(plot.data$CellGroup, levels = celltypesOrder)\n",
    "\n",
    "    #scale by cellgroup\n",
    "    goal_sum <- 1\n",
    "    for(celltype in unique(plot.data$CellGroup)){\n",
    "        goal_sum <- pracma::Lcm(\n",
    "            goal_sum,\n",
    "            sum(plot.data$CellGroup == celltype)\n",
    "        )\n",
    "    }\n",
    "\n",
    "    print(goal_sum)\n",
    "\n",
    "    for(celltype in unique(plot.data$CellGroup)){\n",
    "        index_to_copy <- which(match(plot.data$CellGroup, celltype) == 1)\n",
    "        while(sum(plot.data$CellGroup == celltype) != goal_sum){\n",
    "            for (index in index_to_copy){\n",
    "                    plot.data <- rbind(plot.data, plot.data[rep(index, 1), ])\n",
    "\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "\n",
    "    #make Pattern text Vertical\n",
    "    angles <- rep(0, length(unique(plot.data$Pattern)) + length(unique(plot.data$CellGroup)) + length(unique(plot.data$Signaling)))\n",
    "    for (i in seq(length(unique(plot.data$CellGroup))+1,length(unique(plot.data$CellGroup))+3)){\n",
    "        angles[i] <- 90\n",
    "    }\n",
    "\n",
    "    #make our plot\n",
    "    gg <- ggplot(plot.data,\n",
    "        aes(axis1 = CellGroup, axis2 = Pattern, axis3 = Signaling)) +\n",
    "    geom_flow(width = 1/3, aes.flow = \"forward\", aes(fill = Pattern)) + \n",
    "    scale_x_discrete(limits = c(\"Cell groups\", \"Pattern\", \"Signaling\" )) +\n",
    "    geom_stratum(alpha = 0.8, aes(fill = Pattern),width = 1/3, size=0.1) + \n",
    "    scale_fill_manual(values = color.use) +\n",
    "    geom_text(angle=angles, size = 4, stat = \"stratum\", aes(label = after_stat(stratum))) +\n",
    "    theme_bw()+\n",
    "        theme(legend.position = \"none\",\n",
    "            axis.title = element_blank(),\n",
    "            axis.text.y= element_blank(),\n",
    "            panel.grid.major = element_blank(),\n",
    "            panel.grid.minor  = element_blank(),\n",
    "            panel.border = element_blank(),\n",
    "            axis.ticks = element_blank(),axis.text=element_text(size= 10)) + \n",
    "            theme(plot.title = element_text(hjust = 0.5)) +\n",
    "    ggtitle(main.title)\n",
    "    return(gg)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b569658-83e3-44ee-9a47-9c48a6f9c6a0",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### NMF_and_river"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c0300b97-1507-4197-9223-dfaae356037d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ----------------------------\n",
    "# This function takes in cellchat object and is responsibule for doing all the runs\n",
    "# and making all the figures\n",
    "# ----------------------------\n",
    "NMF_and_river <- function(cellchat_object,study,disease, inhibitory_celltypes, excitatory_celltypes, support_celltypes, parallel='p16') {\n",
    "    for (chosen_seed in seeds){\n",
    "        for (chosen_nrun in nrums){\n",
    "            for (chosen_method in methods){\n",
    "                for (chosen_version in versions){\n",
    "                    try({\n",
    "                    slot.name = \"netP\"\n",
    "                    pattern = \"outgoing\"\n",
    "                    object = cellchat_object\n",
    "                    k = 3 #three patterns\n",
    "\n",
    "                    prob <- methods::slot(object, slot.name)$prob\n",
    "\n",
    "                    data_sender <- apply(prob, c(1,3), sum)\n",
    "                    data_sender = sweep(data_sender, 2L, apply(data_sender, 2, function(x) max(x, na.rm = TRUE)), '/', check.margin = FALSE)\n",
    "                    data0 = as.matrix(data_sender)\n",
    "\n",
    "                    options(warn = -1)\n",
    "                    data <- data0\n",
    "                    data <- data[rowSums(data)!=0,]\n",
    "                    print(\"beginning NMF\")\n",
    "                    outs_NMF <- NMF::nmf(data, rank = 3, method = chosen_method, seed = chosen_seed, nrun=chosen_nrun, .opt=parallel) \n",
    "                    W <- scaleMat(outs_NMF@fit@W, 'r1')\n",
    "                    H <- scaleMat(outs_NMF@fit@H, 'c1')\n",
    "                    print(\"Finished NMF\")\n",
    "                    colnames(W) <- paste0(\"Pattern \", seq(1,ncol(W))); rownames(H) <- paste0(\"Pattern \", seq(1,nrow(H)));\n",
    "\n",
    "                    data_W <- as.data.frame(as.table(W)); colnames(data_W) <- c(\"CellGroup\",\"Pattern\",\"Contribution\")\n",
    "                    data_H <- as.data.frame(as.table(H)); colnames(data_H) <- c(\"Pattern\",\"Signaling\",\"Contribution\")\n",
    "\n",
    "                    res.pattern = list(\"cell\" = data_W, \"signaling\" = data_H)\n",
    "                    methods::slot(object, slot.name)$pattern[[pattern]] <- list(data = data0, pattern = res.pattern)\n",
    "                    \n",
    "                    base_file_name = paste0(\"outgoing\", chosen_nrun, \"_runs_\", chosen_version, \".pdf\")\n",
    "                    rds_name = paste0(\"outgoing\", chosen_nrun, \"_runs_\", chosen_version, \".rds\")\n",
    "                    folder_name = paste(study,disease,sep=\"/\")\n",
    "                    file_name = paste(chosen_method,chosen_seed,base_file_name,sep=\"_\")\n",
    "                    print(\"creating directory\")\n",
    "                    file_dir = paste(folder_name,chosen_method,chosen_seed,sep=\"/\")\n",
    "                    dir.create(file_dir, recursive = TRUE)\n",
    "                    rds_path = paste(folder_name,chosen_method,chosen_seed,rds_name,sep=\"/\")\n",
    "                    file_path = paste(folder_name,chosen_method,chosen_seed,file_name,sep=\"/\")\n",
    "                    # png(filename=file_path)\n",
    "                    colors = c(\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\")\n",
    "                    my_merged_river(object, pattern = \"outgoing\", disease, \n",
    "                                    inhibitory_celltypes=inhibitory_celltypes, \n",
    "                                    excitatory_celltypes=excitatory_celltypes, \n",
    "                                    support_celltypes=support_celltypes)\n",
    "                    print(\"saving fig:\")\n",
    "                    print(file_path)\n",
    "                    ggsave(file_path, device=\"pdf\")\n",
    "                    # saveRDS(object, file =rds_path)\n",
    "                    print(\"done saving fig\")\n",
    "\n",
    "\n",
    "                    base_file_name = paste0(\"ORG_outgoing\", chosen_nrun, \"_runs_\", chosen_version, \".pdf\")\n",
    "                    folder_name = paste(study,disease,sep=\"/\")\n",
    "                    file_name = paste(chosen_method,chosen_seed,base_file_name,sep=\"_\")\n",
    "                    file_path = paste(folder_name,chosen_method,chosen_seed,file_name,sep=\"/\")\n",
    "                    # png(filename=file_path)\n",
    "                    colors = c(\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\")\n",
    "                    my_netAnalysis_river(object, pattern = \"outgoing\",cutoff=0.5,do.order=TRUE)\n",
    "                    print(\"saving fig:\")\n",
    "                    print(file_path)\n",
    "                    ggsave(file_path, device=\"pdf\")\n",
    "                    print(\"done saving fig\")\n",
    "\n",
    "                    slot.name = \"netP\"\n",
    "                    pattern = \"incoming\"\n",
    "                    object = cellchat_object\n",
    "                    k = 3\n",
    "\n",
    "                    prob <- methods::slot(object, slot.name)$prob\n",
    "\n",
    "                    data_sender <- apply(prob, c(1,3), sum)\n",
    "                    data_sender = sweep(data_sender, 2L, apply(data_sender, 2, function(x) max(x, na.rm = TRUE)), '/', check.margin = FALSE)\n",
    "                    data0 = as.matrix(data_sender)\n",
    "\n",
    "                    options(warn = -1)\n",
    "                    data <- data0\n",
    "                    data <- data[rowSums(data)!=0,]\n",
    "\n",
    "                    outs_NMF <- NMF::nmf(data, rank = 3, method = chosen_method, seed = chosen_seed, nrun=chosen_nrun, .opt=parallel) \n",
    "                    W <- scaleMat(outs_NMF@fit@W, 'r1')\n",
    "                    H <- scaleMat(outs_NMF@fit@H, 'c1')\n",
    "\n",
    "                    colnames(W) <- paste0(\"Pattern \", seq(1,ncol(W))); rownames(H) <- paste0(\"Pattern \", seq(1,nrow(H)));\n",
    "\n",
    "                    data_W <- as.data.frame(as.table(W)); colnames(data_W) <- c(\"CellGroup\",\"Pattern\",\"Contribution\")\n",
    "                    data_H <- as.data.frame(as.table(H)); colnames(data_H) <- c(\"Pattern\",\"Signaling\",\"Contribution\")\n",
    "\n",
    "                    res.pattern = list(\"cell\" = data_W, \"signaling\" = data_H)\n",
    "                    methods::slot(object, slot.name)$pattern[[pattern]] <- list(data = data0, pattern = res.pattern)\n",
    "                    \n",
    "                    \n",
    "                    base_file_name = paste0(\"incoming\", chosen_nrun, \"_runs_\", chosen_version, \".pdf\")\n",
    "                    rds_name = paste0(\"incoming\", chosen_nrun, \"_runs_\", chosen_version, \".rds\")\n",
    "                    folder_name = paste(study,disease,sep=\"/\")\n",
    "                    file_name = paste(chosen_method,chosen_seed,base_file_name,sep=\"_\")\n",
    "\n",
    "                    file_dir = paste(folder_name,chosen_method,chosen_seed,sep=\"/\")\n",
    "                    dir.create(file_dir, recursive = TRUE)\n",
    "                    print(\"making directory\")\n",
    "                    file_path = paste(folder_name,chosen_method,chosen_seed,file_name,sep=\"/\")\n",
    "                    rds_path = paste(folder_name,chosen_method,chosen_seed,rds_name,sep=\"/\")\n",
    "                    # png(filename=file_path)\n",
    "                    colors = c(\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\",\"green\")\n",
    "                    my_merged_river(object, pattern = \"incoming\", disease, \n",
    "                                    inhibitory_celltypes=inhibitory_celltypes, \n",
    "                                    excitatory_celltypes=excitatory_celltypes, \n",
    "                                    support_celltypes=support_celltypes)\n",
    "                    print(\"saving fig:\")\n",
    "                    print(file_path)\n",
    "                    ggsave(file_path, device=\"pdf\")\n",
    "                    # saveRDS(object, file =rds_path)\n",
    "                    print(\"done saving fig\")\n",
    "\n",
    "                    base_file_name = paste0(\"ORG_incoming\", chosen_nrun, \"_runs_\", chosen_version, \".pdf\")\n",
    "                    folder_name = paste(study,disease,sep=\"/\")\n",
    "                    file_name = paste(chosen_method,chosen_seed,base_file_name,sep=\"_\")\n",
    "                    file_path = paste(folder_name,chosen_method,chosen_seed,file_name,sep=\"/\")\n",
    "                    # png(filename=file_path)\n",
    "                    my_netAnalysis_river(object, pattern = \"incoming\",cutoff=0.5,do.order=TRUE)\n",
    "                    print(\"saving fig:\")\n",
    "                    print(file_path)\n",
    "                    ggsave(file_path, device=\"pdf\")\n",
    "                    print(\"done saving fig\")\n",
    "\n",
    "                    })\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f5a449d-f6e9-48f5-ab4b-d4d2997c7409",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
